### PR TITLE
[registrar] Add support for specifying that a protocol changed informal status in a certain SDK. Fixes #43780 and #48311.

### PR DESCRIPTION
**Pull request by rolfbjarne**
Add support for specifying that an informal protocol became a formal protocol
(or the reverse) in a certain SDK version, so that the static registrar can
generate the correct code based on the SDK being built with.

This also fixes a series of compiler warnings when using the static registrar:

    In file included from Xamarin.Mac.registrar.full.i386.m:1:
    ./Xamarin.Mac.registrar.full.i386.h:374:11: warning: duplicate protocol definition of 'CALayerDelegate' is ignored
    @protocol CALayerDelegate
              ^
    /Applications/Xcode83.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk/System/Library/Frameworks/QuartzCore.framework/Headers/CALayer.h:798:11: note: previous definition is here
    @protocol CALayerDelegate <NSObject>
              ^
    In file included from Xamarin.Mac.registrar.full.i386.m:1:
    ./Xamarin.Mac.registrar.full.i386.h:824:11: warning: duplicate protocol definition of 'WebDownloadDelegate' is ignored
    @protocol WebDownloadDelegate
              ^
    /Applications/Xcode83.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk/System/Library/Frameworks/WebKit.framework/Headers/WebDownload.h:60:11: note: previous definition is here
    @protocol WebDownloadDelegate <NSURLDownloadDelegate>
              ^
    In file included from Xamarin.Mac.registrar.full.i386.m:1:
    ./Xamarin.Mac.registrar.full.i386.h:851:11: warning: duplicate protocol definition of 'WebFrameLoadDelegate' is ignored
    @protocol WebFrameLoadDelegate
              ^
    /Applications/Xcode83.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk/System/Library/Frameworks/WebKit.framework/Headers/WebFrameLoadDelegate.h:51:11: note: previous definition is here
    @protocol WebFrameLoadDelegate <NSObject>
              ^
    In file included from Xamarin.Mac.registrar.full.i386.m:1:
    ./Xamarin.Mac.registrar.full.i386.h:866:11: warning: duplicate protocol definition of 'WebPolicyDelegate' is ignored
    @protocol WebPolicyDelegate
              ^
    /Applications/Xcode83.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk/System/Library/Frameworks/WebKit.framework/Headers/WebPolicyDelegate.h:138:11: note: previous definition is here
    @protocol WebPolicyDelegate <NSObject>
              ^
    In file included from Xamarin.Mac.registrar.full.i386.m:1:
    ./Xamarin.Mac.registrar.full.i386.h:869:11: warning: duplicate protocol definition of 'WebResourceLoadDelegate' is ignored
    @protocol WebResourceLoadDelegate
              ^
    /Applications/Xcode83.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk/System/Library/Frameworks/WebKit.framework/Headers/WebResourceLoadDelegate.h:46:11: note: previous definition is here
    @protocol WebResourceLoadDelegate <NSObject>
              ^
    In file included from Xamarin.Mac.registrar.full.i386.m:1:
    ./Xamarin.Mac.registrar.full.i386.h:872:11: warning: duplicate protocol definition of 'WebUIDelegate' is ignored
    @protocol WebUIDelegate
              ^
    /Applications/Xcode83.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk/System/Library/Frameworks/WebKit.framework/Headers/WebUIDelegate.h:153:11: note: previous definition is here
    @protocol WebUIDelegate <NSObject>
              ^

https://bugzilla.xamarin.com/show_bug.cgi?id=43780
https://bugzilla.xamarin.com/show_bug.cgi?id=48311

### CODE REVIEW

File: src/Foundation/ProtocolAttribute.cs [hunk 1]
```diff
@@ -34,6 +34,21 @@ public sealed class ProtocolAttribute : Attribute {
 		public Type WrapperType { get; set; }
 		public string Name { get; set; }
 		public bool IsInformal { get; set; }
+		// In which SDK version this protocol switched from (or to, depending on the current IsInformal value) being informal.
+		// If IsInformal = true, then this protocol switched to being informal in the specified SDK version
+		// If IsInformal = false, then this protocol switched to a real protocol in the specified SDK version
+		// Version is not a valid type for attributes, so we're using an array of ints.
+		int[] informal_switch;
+		public int [] InformalSwitch {
```
Comment (spouliot) [inline src/Foundation/ProtocolAttribute.cs]: A string would be more costly to parse right?

File: src/coreanimation.cs [hunk 2]
```diff
@@ -858,7 +858,15 @@ interface CATextLayer {
 
 	[BaseType (typeof (NSObject))]
 	[Model]
-	[Protocol (IsInformal = true)] // not informal as of iOS 10+, but removing the IsInformal value breaks when building with older SDKs (see bug #43585).
+#if IOS || TVOS
+	[Protocol (InformalSwitch = new int [] { 10, 0 })]
```
Comment (spouliot) [inline src/coreanimation.cs]: Are there cases where protocols became informal, ie after being formal ?
Reading this the original state is unclear

File: src/Foundation/ProtocolAttribute.cs (hunk 1 above)
Comment (rolfbjarne) [inline src/Foundation/ProtocolAttribute.cs]: I didn't consider using a string, but I don't think it's much of a difference cost-wise (and in any case this is rarely needed).

I checked whether the metadata is bigger for strings or int arrays, and couldn't find any difference at all up to 100.000 attributes.

So I guess it comes down to readability, and I think strings are slightly more readable, so I'll change it.

File: src/coreanimation.cs (hunk 2 above)
Comment (rolfbjarne) [inline src/coreanimation.cs]: No, I haven't seen any cases where  protocols have become informal after being formal. I'll simplify it to only handle the other case, to make the code clearer.

File: src/Foundation/ProtocolAttribute.cs (hunk 1 above)
Comment (rolfbjarne) [inline src/Foundation/ProtocolAttribute.cs]: @spouliot this should be fixed now

**Comment by rolfbjarne** [comment]
Yeah, `FormalSince` sounds better, also a bit less ambiguous than `InformalUntil`.

**Pull request merged**